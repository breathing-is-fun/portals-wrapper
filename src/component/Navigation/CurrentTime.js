import React, { Component } from 'react';

import moment from 'moment';

export default class CurrentTime extends Component {
  constructor (props) {
    super(props);

    this.state = {
      time: moment().format('HH:mm:ss')
    };
  }

  componentDidMount = () => {
    this.getCurrentTime();
  }

  getCurrentTime = () => {
    setInterval(() => {
      this.setState({ time: moment().format('HH:mm:ss') });
    }, 1000);
  }

  render = () => {
    const { time } = this.state;

    return (
      <div className='CurrentTime'>{time}</div>
    );
  }
}