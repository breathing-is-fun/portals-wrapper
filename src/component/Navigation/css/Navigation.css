.Navigation .content {
  height: 65px;
  background: #2e7df3;
  box-shadow: 0px 0px 9px 0px rgba(39, 93, 214, 0.39);
}

.Navigation .content-sm {
  height: 40px;
  background: #2e7df3;
  box-shadow: 0px 0px 9px 0px rgba(39, 93, 214, 0.39);
}

.Navigation .children-wrapper {
  overflow-x: hidden;
  overflow-y: auto;
  background-color: #ffffff;
  box-shadow: 0px 0px 29px 0px rgba(93, 106, 113, 0.12);
  border-radius: 2px;
}

.Navigation .content-wrapper {
  font-size: 28px;
  color: #ffffff;
  float: left;
  height: 100%;
  line-height: 65px;
  text-align: center;
}

.Navigation .sm {
  font-size: 20px;
  color: #ffffff;
  float: left;
  height: 100%;
  line-height: 40px;
  text-align: center;
}

.Navigation .content-switch {
  float: right;
  margin-right: 20px;
  color: #FFF;
}

.Navigation .content-space {
  width: 1px;
  height: 100%;
  background: #2372e7;
  float: left;
}

.Navigation .content-time {
  font-size: 34px;
  color: #ffffff;
}

.Navigation .content-time-detail {
  height: 24px;
  font-size: 13px;
  line-height: 33px;
  color: #ffffff;
}

.Navigation .content-wrapper .droplink {
  color: #FFF;
  background-color: transparent;
  text-decoration: none;
  outline: none;
  cursor: pointer;
}

.Navigation .content-wrapper .droplink .droplink-icon {
  font-size: 18px;
  transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
}