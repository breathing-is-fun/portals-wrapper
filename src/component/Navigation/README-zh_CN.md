[English](./README.md) | 简体中文

# 简介

- 类似于 `banner` 的头部

- 用作切换[套餐](http://www.baidu.com)

- 基于 [antd的Dropdown](https://ant.design/components/dropdown-cn/)
  
# API  

| 名称 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| clock | 是否显示时间 | boolean | - |
| title | 左上角的标题 | string | - |
| onClick | 点击菜单项回调 | function (item: object) | - |
| datas | 菜单数据，只有右边这俩参数是必须的，其他的你传进来也没问题，反正最后会把这个 `object` 传出来 | [{ text: string, id: string, url: string }] | [] |

# 示例

``` jsx
    import Navigation from '.'

    render = () => {
        return (
            <Navigation
                onClick={ ({ id }) => console.log(id) }
                datas={ [] }
                clock
            >
                test
            </Navigation>
        )
    }
```

![img](../../../docs/img/Navigation.gif)