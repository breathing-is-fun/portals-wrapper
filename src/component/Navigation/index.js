/*
 * @Author: zy9@github.com/zy410419243
 * @Date: 2018-10-15 15:47:19
 * @Last Modified by: zy9
 * @Last Modified time: 2018-10-18 16:38:28
 */
import React, { Component } from 'react';

import moment from 'moment';
moment.locale('zh-cn');

import './css/Navigation.css';

export default class Navigation extends Component {
	constructor (props) {
		super(props);

		this.state = {
			time: moment().format('HH:mm:ss'),
		};
	}

    componentDidMount = () => {
    	this.getCurrentTime();
    }

    getCurrentTime = () => {
    	setInterval(() => {
    		this.setState({ time: moment().format('HH:mm:ss') });
    	}, 1000);
    }

    render = () => {
    	const { children } = this.props;
    	const { time } = this.state;

    	// 分割线
    	const space = <div className='content-space' />;

    	return (
    		<div className='Navigation'>
    			<div className='content'>
    				<div className='content-wrapper' style={{ width: '15%' }}>浙江省水利综合门户</div>

    				{/* { space } */}
    				<div className='content-wrapper' style={{ width: '15%' }}>
    					<div className='content-time'>{ time }</div>
    				</div>

    				<div className='content-wrapper' style={{ marginLeft: '-3.5%' }}>
    					<div className='content-time-detail'>{ moment().format('YYYY-MM-DD') }</div>
    					<div className='content-time-detail'>{ moment().format('dddd') }</div>
    				</div>
    			</div>

    			{ children }
    		</div>
    	);
    }
}