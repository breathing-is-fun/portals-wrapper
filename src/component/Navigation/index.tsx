import React, { Component } from 'react';
import classNames from 'classnames';
import { Menu, Dropdown, Icon } from 'antd';
import moment from 'moment';
moment.locale('zh-cn');
import './css/Navigation.css';

export interface NavigationProps {
  title?: string;
  datas?: Array<any>;
  clock?: boolean;
  size?: string | number;
  display?: boolean;
  onClick?: (item: any) => void;
}
export interface NavigationState {
  CurrentTime: any;
  iconRotate: number | string;
  dropDownVisible: boolean;
  clientHeight: number;
  currentSelect: string;
}

export default class Navigation extends Component<
  NavigationProps,
  NavigationState
> {
  static defaultProps = {
    datas: [],
  };

  constructor(props: any) {
    super(props);

    this.state = {
      CurrentTime: null,
      iconRotate: 0,
      dropDownVisible: false,
      clientHeight: document.body.clientHeight,
      currentSelect: '',
    };
  }

  static getDerivedStateFromProps = (
    props: NavigationProps,
    state: NavigationState,
  ) => {
    const { datas = [] } = props;
    const { currentSelect } = state;
    const newState: any = {};

    if (!currentSelect && datas.length != 0) {
      newState.currentSelect = datas[0].title;
    }
    return newState;
  };

  componentDidMount = () => {
    if ((window as any).SCTool) {
      SCTool.RegisterResizeDispatcher = {
        key: 'Navigation',
        onResize: (item: any) => {
          this.setState({ clientHeight: item.clientHeight });
        },
      };
    }

    this.props.clock && this.loadCurrentTime();
  };

  loadCurrentTime = () => {
    import('./CurrentTime').then(CurrentTime => {
      this.setState({ CurrentTime: CurrentTime.default });
    });
  };

  handleIconType = (dropDownVisible: boolean) => {
    let { iconRotate } = this.state;
    iconRotate = dropDownVisible ? '180' : '0';

    this.setState({
      iconRotate,
      dropDownVisible,
    });
  };

  handleMenuItemClick = (item: any) => {
    const { onClick } = this.props;
    const { title } = item;

    this.setState({ currentSelect: title }, () => {
      onClick && onClick(item);
    });
  };

  render = () => {
    const { children, datas = [], clock, title, size, display } = this.props;
    const {
      CurrentTime,
      dropDownVisible,
      iconRotate,
      clientHeight,
      currentSelect,
    } = this.state;

    const date = (
      <div
        className={classNames({
          'content-wrapper': true,
          sm: size == 'sm',
        })}
      >
        <div
          className={classNames({
            'content-time-detail': true,
            sm: size == 'sm',
          })}
        >
          {moment().format('YYYY-MM-DD')}
        </div>
        {size != 'sm' && (
          <div
            className={classNames({
              'content-time-detail': true,
              sm: size == 'sm',
            })}
          >
            {moment().format('dddd')}
          </div>
        )}
      </div>
    );

    const switchMeal = (
      <Menu>
        {datas.map(item => {
          const { title, id } = item;

          return (
            <Menu.Item
              key={id}
              onClick={(e: any) => this.handleMenuItemClick(item)}
            >
              <a rel="noopener noreferrer">{title}</a>
            </Menu.Item>
          );
        })}
      </Menu>
    );

    return (
      <div className="Navigation">
        <div
          className={classNames({
            content: size == '',
            'content-sm': size == 'sm',
          })}
          style={{ display: display ? '' : 'none' }}
        >
          <div
            className={classNames({
              'content-wrapper': true,
              sm: size == 'sm',
            })}
            style={{ width: 300 }}
          >
            {title}
          </div>

          <div
            className={classNames({
              'content-wrapper': true,
              sm: size == 'sm',
            })}
            style={{ width: 300 }}
          >
            <div
              className={classNames({
                'content-time': size == '',
                sm: size == 'sm',
              })}
            >
              {CurrentTime && <CurrentTime />}
            </div>
          </div>

          {clock && date}

          {datas && datas.length != 0 && (
            <div
              className={classNames({
                'content-wrapper': true,
                'content-switch': true,
                sm: size == 'sm',
              })}
            >
              <Dropdown
                overlay={switchMeal}
                trigger={['hover', 'click']}
                onVisibleChange={this.handleIconType}
                visible={dropDownVisible}
              >
                <span className="droplink">
                  {currentSelect}
                  <Icon
                    type="down"
                    style={{
                      transform: `rotate(${iconRotate}deg)`,
                      marginLeft: 10,
                    }}
                    className="droplink-icon"
                  />
                </span>
              </Dropdown>
            </div>
          )}
        </div>

        <div
          className="children-wrapper"
          style={{ height: clientHeight - 65 + 'px' }}
        >
          {children}
        </div>
      </div>
    );
  };
}
