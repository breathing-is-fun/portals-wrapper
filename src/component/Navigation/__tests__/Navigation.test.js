/*
 * @Author: zy9@github.com/zy410419243
 * @Date: 2018-09-21 09:03:47
 * @Last Modified by: zy9
 * @Last Modified time: 2018-10-21 12:57:35
 */
import React from 'react';
import { render, mount, shallow } from 'enzyme';
import Navigation from '..';

describe('Navigation', () => {
	it('render correctly', () => {
		const wrapper = render(
			<Navigation>test</Navigation>
		);

		expect(wrapper).toMatchSnapshot();
	});

	// it('when operate drawer is clicked, onChange should to be called', () => {
	// 	const onChange = jest.fn();
	// 	const wrapper = mount(
	// 		<Navigation visible={ true } onChange={ onChange } direction='right'>test</Navigation>
	// 	);

	// 	wrapper.find('.operate-drawer').simulate('click');

	// 	expect(onChange).toBeCalled();
	// });
});