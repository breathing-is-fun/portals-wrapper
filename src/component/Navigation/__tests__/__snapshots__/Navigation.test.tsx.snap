// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation render correctly 1`] = `
<Navigation
  datas={Array []}
>
  <div
    className="Navigation"
  >
    <div
      className=""
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div
        className="content-wrapper"
        style={
          Object {
            "width": 300,
          }
        }
      />
      <div
        className="content-wrapper"
        style={
          Object {
            "width": 300,
          }
        }
      >
        <div
          className=""
        />
      </div>
    </div>
    <div
      className="children-wrapper"
      style={
        Object {
          "height": "-65px",
        }
      }
    >
      test
    </div>
  </div>
</Navigation>
`;
